tensorflow_model_optimization/python/core/sparsity_tf2/pruner_test.py

```diff
@@ -0,0 +1,493 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for the key functions in pruner library."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# import g3
+
+from absl.testing import parameterized
+import numpy as np
+import tensorflow as tf
+
+# TODO(b/139939526): move to public API.
+from tensorflow.python.keras import keras_parameterized
+from tensorflow_model_optimization.python.core.keras import compat
+from tensorflow_model_optimization.python.core.sparsity_tf2 import pruner
+from tensorflow_model_optimization.python.core.sparsity.keras import pruning_schedule
+from tensorflow_model_optimization.python.core.sparsity.keras import pruning_utils
+
+K = tf.keras.backend
+dtypes = tf.dtypes
+test = tf.test
+
+
+def assign_add(ref, value):
+  if hasattr(tf, "assign_add"):
+    return tf.assign_add(ref, value)
+  else:
+    return ref.assign_add(value)
+
+def get_lth_sparsity(save_round, n_rounds, target_sparsity, end_epoch):
+  """
+  save_round: when to save the weight; 1 onwards
+  n_rounds: number of pruning cycles to do
+  target_sparsity: the sparsity percentage to achieve by the end of the iteration
+  end_epoch: total number of epochs to train for where pruning is eligible
+
+  Returns:
+    percent to prune to after each cyle/round
+  """
+  # no pruning until weights are saved TODO: off by one??
+  frequency = tf.math.floordiv(end_epoch - save_round + 1, n_rounds) # range(0, end, freq)
+  prune_pcent_per_round = tf.math.pow(target_sparsity, tf.math.divide(1, n_rounds))
+  return tf.cast(frequency, tf.int64), prune_pcent_per_round
+
+def make_pruning_schedule(target_sparsity, begin, end, freq):
+  return pruning_schedule.ConstantSparsity(target_sparsity, begin, end, freq)
+
+def sample_noise(x, mu=0, sigma=1.):
+  dist = tf.random.normal(loc=mu,  sccale=sigma)
+  return dist.prob(x)
+
+class PruningTest(test.TestCase, parameterized.TestCase):
+
+  def setUp(self):
+    super(PruningTest, self).setUp()
+    self.block_size = (1, 1)
+    self.block_pooling_type = "AVG"
+    self.target_sparsity = 0.5
+    self.constant_sparsity = pruning_schedule.ConstantSparsity(self.target_sparsity, 0, 100, 1)
+    self.save_init = tf.Variable(0)
+    self.save_itr_10 = 10
+    # dummy optimizer
+    self.dummy_optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
+    self.grad = lambda shape: tf.ones_like(shape)
+
+  # Variable initialization outside of setUp() is needed for compatibility with
+  # run_all_keras_modes.
+  #
+  # setUp() lies outside of the "eager scope" that wraps the test cases
+  # themselves, resulting in initializing graph tensors instead of eager
+  # tensors when testing eager execution.
+  def initialize(self):
```

Comment:
Delete this & all references to this or `self.global_step` or `train_step_fn`

Always grab the optimizer's `opt.iterations` if you need the step variable